### ISSUE TITLE
Upload Issue When Providing No Sources With Direct URL

### ISSUE THREAD

**Issue by DonovanDMC** [zwagoth/e621ng#337]
Uploading with `direct_url`, and no sources results in an error, as there is an attempt to add to sources, which does not exist
even using an empty value results in the same error (e.g. `upload[source]=`, [ends up nil](https://butts-are.cool/WindowsTerminal_10-19-2021_20-09-52.png) (at least it does when I try with nodejs' http(s)))

The resulting error is ([code link](https://github.com/zwagoth/e621ng/blob/a28466bdaf2b6c3826c9d0914387c9305983397d/app/models/upload.rb#L334)):
```
NoMethodError: undefined method `+' for nil:NilClass
  app/models/upload.rb:334:in `fixup_source'
  app/logical/upload_service.rb:13:in `start!'
  app/controllers/uploads_controller.rb:43:in `block in create'
  app/controllers/uploads_controller.rb:41:in `create'
```

This happened on a bare bones development instance with no modifications whatsoever.

**Comment by Earlopain** [comment]
I'm not sure what the problem here is, there must be something you are doing wrong since the uploader has no problems doing it.
Maybe this code snippet will help you, it's ripped straight from the uploaders submit function. I can reproduce your error message when `upload[source]` is not appended.
```js
data = new FormData();

data.append('upload[direct_url]', "whatever_url_you_want.png");
data.append('upload[tag_string]', "a b c d e f g");
data.append('upload[rating]', "s");
data.append('upload[source]', "");
data.append('upload[description]', "");
data.append('upload[parent_id]', "");

jQuery.ajax('/uploads.json', {
    contentType: false,
    processData: false,
    method: 'POST',
    type: 'POST',
    data: data
});
```

**Comment by DonovanDMC** [comment]
Something I was doing was causing the issue with empty sources, but the docs don't state source as being required, and logically with a lot of posts having no sources it shouldn't be required 

I get that it should be provided as empty, but the documentation doesn't say that, and having a very not helpful error when not providing source seems counter intuitive 

I made minimum reprod scripts in [[ts](https://pastebin.com/p894iHvC)] and [[js](https://pastebin.com/5BM1LWdQ)]

**Comment by Earlopain** [comment]
It is documented. On the api page is says `There are only four mandatory fields: you need to supply the file (either through a multipart form or through a source URL), the tags, a source (even if blank), and the rating.`

https://e621.net/help/api#posts_create

**Comment by DonovanDMC** [comment]
I'm blind then, neat

still seems wrong to have an internal error rather than a check for source not being present
uploading with a file still works just fine because it doesn't try to add anything to the sources

**Comment by Earlopain** [comment]
I agree, a better error message wouldn't hurt. I'll leave this open and get to it at a later date

**Comment by DonovanDMC** [comment]
I think I've finally gotten down to my problem, take this body for example (importing posts from e6 to local)
![](https://butts-are.cool/screenshot_11-26-2021_09-12-21.png)

I think this here is my problem
![](https://butts-are.cool/screenshot_11-26-2021_09-13-05.png)

it seems like I'm throwing away the equals sign somewhere and that's causing ruby's parser to pull out nil values

Yep, that was my problem -- so it's an issue of badly formed requests (this here passed back a db error)
![](https://butts-are.cool/screenshot_11-26-2021_09-19-09.png)

**Comment by Earlopain** [comment]
Seems to be acidentally fixed by #357, more specifically 65fed7a800ca5e3adc9f22255cc136ccc290213a which just removes the function altogether. When this eventually gets merged providing a empty source isn't mandatory anymore
